### DISCUSSION TITLE
Building bottles for arm64

### DISCUSSION THREAD

**Discussion by ValentinTrinque** [Homebrew/discussions#2896] [Tap maintenance and Homebrew development]
I am trying to find a way to build arm64 bottles on a tap, using the Github Actions generated by `brew tap-new` that uses `brew test-bot`, but I couldn't find out how. I had a look to the homebrew-core repository to find out how it's doing it, but got lost.

I tried to keep it simple.

My formula is targeting a golang project, so [I can cross-compile by setting the GOARCH env var](https://github.com/vegaprotocol/homebrew-vega/blob/02885f6ff733b2ea93228455f80791fd8e5ba957/.github/workflows/tests.yml#L19).

At this point, I assume the project is correctly compiled for arm64.

However, the default file name of the bottle, when creating it, resolves on the underlaying CPU architecture (x86-64) [(job)](https://github.com/vegaprotocol/homebrew-vega/runs/5012450038?check_suite_focus=true):
```sh
122  ==> brew bottle --verbose --json vegaprotocol/vega/vegawallet --root-url=https://github.com/vegaprotocol/homebrew-vega/releases/download/vegawallet-0.11.0
123  ==> Determining vegaprotocol/vega/vegawallet bottle rebuild...
124  ==> Bottling vegawallet--0.11.0.x86_64_linux.bottle.tar.gz...
125  tar --create --numeric-owner --file /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/vegaprotocol/homebrew/vega/vegawallet--0.11.0.x86_64_linux.bottle.tar vegawallet/0.11.0
126  ==> Detecting if vegawallet--0.11.0.x86_64_linux.bottle.tar.gz is relocatable...
127  ./vegawallet--0.11.0.x86_64_linux.bottle.tar.gz
128    bottle do
129      root_url "https://github.com/vegaprotocol/homebrew-vega/releases/download/vegawallet-0.11.0"
130      sha256 cellar: :any_skip_relocation, x86_64_linux: "aa636dfcfce537df032de69eebab3bd463222ffe49aabe64dc523379be2384ee"
131    end
132  Writing vegawallet--0.11.0.x86_64_linux.bottle.json
133  ==> brew bottle --merge --write --no-commit ./vegawallet--0.11.0.x86_64_linux.bottle.json
134  ==> vegaprotocol/vega/vegawallet
135    bottle do
136      root_url "https://github.com/vegaprotocol/homebrew-vega/releases/download/vegawallet-0.11.0"
137      sha256 cellar: :any_skip_relocation, x86_64_linux: "aa636dfcfce537df032de69eebab3bd463222ffe49aabe64dc523379be2384ee"
138    end
```

Is there a way to make brew test-bot resolve to arm64 during the bottling ?

Is this possible to reproduce what homebrew-core is doing to auto create the arm64 bottles ?

Supporting this out of the box could be nice by the Github Action.

Reference:
* [Homebrew tap with bottles uploaded to GitHub Releases - 18 November 2020](https://brew.sh/2020/11/18/homebrew-tap-with-bottles-uploaded-to-github-releases/)
* [Vega Protocol tap - what I try go get working for arm64](https://github.com/vegaprotocol/homebrew-vega)
* [The formula](https://github.com/vegaprotocol/homebrew-vega/blob/main/Formula/vegawallet.rb)
* [The Github Action](https://github.com/vegaprotocol/homebrew-vega/blob/main/.github/workflows/tests.yml)

**Comment by carlocab** [comment]
Not out of the box. One way might be to patch the call to `uname -m` [here](https://github.com/Homebrew/brew/blob/f05e57f41421315247c7566c35b47807972b3d0a/Library/Homebrew/brew.sh#L8) and replace it with `arm64` (or `aarch64`, if that's the output of `uname -m` on ARM64 Linux).

Note sure if this will work, but I don't see a reason why it shouldn't. If you don't want to patch `brew`, the only way is probably to build bottles on an ARM64 machine (or perhaps an ARM64 VM).

**Reply by tedim52** [reply]
@carlocab Thank you for response here. 

I'm having a very similar issue - I do not have an `arm64` machine to run `brew bottle...` on, however, my [formula](https://github.com/kurtosis-tech/homebrew-tap/blob/main/kurtosis-cli.rb) works in a way that based on the OS/architecture, the script will install a specific binary from a url. Thus, it shouldn't be dependent on the underlying machine. 

Is there anyway to use `brew bottle...` such that the bottle is taking the binary from the OS/architecture specified in the formula (pointing to a binary already built for the desired OS/architecture) as opposed to resolving the OS/arch from the underlying machine? If not, is there a way to patch this in brew, or openess to solve this problem?

**Reply by SMillerDev** [reply]
That sounds like it should be a cask instead. 

Bottles are prebuilt binaries so bottling prebuilt binaries is only causing problems here.

**Reply by tedim52** [reply]
Thank you for your response @SMillerDev! Will definitely look into how cask can solve our problem.

For context, currently, `goreleaser` is publishing the homebrew formula when we release, and works in such a way that the formula points to prebuilt goreleaser binaries hosted on a repo. We want to use bottles to enable installation of older versions of our binary, by hosting the bottles on our tap repo. Ideally, we wouldn't be building two different binaries (goreleaser binaries AND bottles) or bottling a prebuilt binary but bottles give users an easy installation step via `brew install -force-bottle <bottle> tar` whereas the `goreleaser` binaries do not (from my understanding). LMK if you have any ideas/suggestions/or advice here?

**Reply by tedim52** [reply]
For additional context, we'd like to build bottles for `x86_64` AND `arm64` machines and our problem could be solved right now if github actions supported `arm64` runners, but they currently do not support that - and we'd ideally not have to create our own self hosted `arm64` runner if possible.

**Reply by christoffer-eide** [reply]
There is a demand for GH hosted arm64 runners. But there is no response from GH regarding this.

One of the tips is to use buildjet for hosted arm64 runners:
https://github.com/actions/runner-images/issues/5631#issuecomment-1415084822

**Reply by heaths** [reply]
arm64 runners are GA: https://github.blog/changelog/2024-09-03-github-actions-arm64-linux-and-windows-runners-are-now-generally-available/

**Reply by heaths** [reply]
Adding in the `ubuntu-22.04-arm` image, I've been able to make some headway: https://github.com/heaths/homebrew-tap/pull/7. I have to disable running `brew test-bot --only-setup`, though, because `brew doctor` fails with a warning due ti linux-arm64 being a tier 2. Is there any other way to disable this apart from disabling the entire step? I've looked through `brew test-bot --help` and documentation and found no way to disable that warning, but I imagine there must be some way.

**Reply by ReenigneArcher** [reply]
The `ubuntu-22.04-arm` image is working for me, but I had to add a hack to solve a ruby problem. https://github.com/orgs/Homebrew/discussions/6615#discussioncomment-15561564

I didn't have to disable `brew test-bot --only-setup`.

Here's my workflow for reference: https://github.com/LizardByte/homebrew-homebrew/blob/3ba6e203fefea6f26016fb801c3ab1c26521b85e/.github/workflows/tests.yml#L326-L340

**Reply by heaths** [reply]
> I didn't have to disable brew test-bot --only-setup.

Because they recently fixed it. It was broken on ARM as of my post last spring.
